### PR TITLE
[exporter/loadbalancing] Adding AWS Cloud Map for Load Balancer Exporter discovery

### PR DESCRIPTION
**Pull request by andretong**
**Description:** 
Implementation of [27241](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/27241)

Adding AWS Service discovery system (CloudMap) support to resolve the Collector's Backend.
This would allow users to use this exporter when there is the case of using ECS over EKS in an AWS infrastructure.

**Link to tracking Issue:** <Issue number if applicable>

**Testing:** <Describe what testing was performed and which tests were added.>

Setting the following configuration with the proper values, we can start the Load Balancer which will connect to the AWS CloudMap service and start sending traces to the available collectors.

```yaml
exporters:
  loadbalancing:
    protocol:
      otlp:
        timeout: 3s
    resolver:
      awsCloudMap:
        namespace: <aws-namespace>
        serviceName: <aws-otel-col-service-name>
        interval: <interval-in-seconds>
```

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

File: exporter/loadbalancingexporter/README.md
Comment (jpkrohling) [reply exporter/loadbalancingexporter/README.md:56]: I think it makes sense to return an error, and users should have this expectation as well. Can you please change this line in the doc to reflect what the code does?

File: exporter/loadbalancingexporter/README.md
Comment (jpkrohling) [inline exporter/loadbalancingexporter/README.md:70]: Is 1s a reasonable value here? If the default interval is 30s, the timeout can certainly be at around 5s, no?

File: exporter/loadbalancingexporter/README.md
Comment (jpkrohling) [inline exporter/loadbalancingexporter/README.md:66]: I know nothing about AWS Cloud Map, is there anything users have to be aware before configuring it?

File: exporter/loadbalancingexporter/README.md
Comment (jpkrohling) [inline exporter/loadbalancingexporter/README.md:71]: There seems to be a `healthStatus` property as well. Should it be documented?

File: exporter/loadbalancingexporter/loadbalancer.go
Comment (jpkrohling) [inline exporter/loadbalancingexporter/loadbalancer.go:55]: This works, and I like that it provides a short-circuit feature to the function. But I propose two alternatives:

1) externalize this into a `func checkMultipleResolversProvided(Config) error`, and call this function here. It should make your code easier to test and leave the existing function more readable.
2) alternatively, you can create a boolean `configured`, and set it to true whenever a resolver is not nil. On the next resolver, you check if configured is true and return the errMultipleResolversProvided error. You won't have the fail-fast behavior, though.

File: exporter/loadbalancingexporter/loadbalancer.go
Comment (jpkrohling) [inline exporter/loadbalancingexporter/loadbalancer.go:105]: For more than 3 parameters, I think I would prefer a new struct with the resolver settings (or reuse the one from the config?)

File: exporter/loadbalancingexporter/resolver_aws_cloudmap.go
Comment (jpkrohling) [inline exporter/loadbalancingexporter/resolver_aws_cloudmap.go:69]: Should this be a TODO? Perhaps it can be a new context with a deadline?

File: exporter/loadbalancingexporter/resolver_aws_cloudmap.go
Comment (jpkrohling) [inline exporter/loadbalancingexporter/resolver_aws_cloudmap.go:71]: Don't throw fatals here, return errors and let it be handled by the caller.

File: exporter/loadbalancingexporter/resolver_aws_cloudmap.go
Comment (jpkrohling) [inline exporter/loadbalancingexporter/resolver_aws_cloudmap.go:77]: It might be good to do a nil check and a len check:

```
if namespaceName == nil || len(*namespaceName) == 0 {
  ...
}
```

Otherwise, you might have a runtime panic if namespaceName is nil. Same for serviceName

File: exporter/loadbalancingexporter/resolver_aws_cloudmap.go
Comment (jpkrohling) [inline exporter/loadbalancingexporter/resolver_aws_cloudmap.go:111]:
```suggestion
		r.logger.Warn("failed initial resolve", zap.Error(err))
```

File: exporter/loadbalancingexporter/resolver_aws_cloudmap.go
Comment (jpkrohling) [inline exporter/loadbalancingexporter/resolver_aws_cloudmap.go:173]: I think this deserves a better message.

File: exporter/loadbalancingexporter/resolver_aws_cloudmap.go
Comment (jpkrohling) [inline exporter/loadbalancingexporter/resolver_aws_cloudmap.go:186]:
```suggestion
		r.logger.Debug("resolved instance",
```

**Review by jpkrohling — APPROVED** [review]
This looks great! Thank you for your PR. @Aneurysm9, @bryan-aguilar, would someone from AWS be available to review this PR?

File: exporter/loadbalancingexporter/README.md
Comment (andretong) [reply exporter/loadbalancingexporter/README.md:71]: For this property, in this initial iteration, I'd rather to leave this config as default since we filter only for Healthy instances

Reference: https://docs.aws.amazon.com/cloud-map/latest/api/API_GetInstancesHealthStatus.html

File: exporter/loadbalancingexporter/README.md
Comment (andretong) [reply exporter/loadbalancingexporter/README.md]: Yes, makes all the sense, I've updated this property and documentation

File: exporter/loadbalancingexporter/README.md
Comment (andretong) [reply exporter/loadbalancingexporter/README.md:66]: The only relevant topic I can think of is that usage of the SDK hits the API every polling interval, so users have to take into account the AWS API Rate limits in their configuration.

**Comment by 0x006EA1E5** [comment]
Hey, this looks great.  Can we expect this to be available in the next release?

File: exporter/loadbalancingexporter/loadbalancer.go
Comment (jpkrohling) [inline exporter/loadbalancingexporter/loadbalancer.go:212]: I'm not entirely comfortable with this change in this PR. I think it has to be on its own PR.

Comment (andretong) [reply exporter/loadbalancingexporter/loadbalancer.go]: Sure, I'll raise a new PR once this is merge with the increment.

**Comment by jpkrohling** [comment]
@andretong, once this is ready for another review, let me know but note that I'm unable to do so this year. I'm back on Jan 15, please ping me again on Jan 18 if I haven't reviewed this here by then.

**Comment by andretong** [comment]
> @andretong, once this is ready for another review, let me know but note that I'm unable to do so this year. I'm back on Jan 15, please ping me again on Jan 18 if I haven't reviewed this here by then.

@jpkrohling thanks for the notice.
All the suggestions were applied and waiting for a review, looking forward to get back with you regarding this feature.

File: exporter/loadbalancingexporter/README.md
Comment (jpkrohling) [reply exporter/loadbalancingexporter/README.md:67]: Can this comment be added to the readme? Or is it so obvious that it doesn't have to?

File: exporter/loadbalancingexporter/README.md
Comment (jpkrohling) [reply exporter/loadbalancingexporter/README.md:72]: I might be missing something, but the value the user is providing is indeed being used, and therefore, deserves to be documented.

File: exporter/loadbalancingexporter/resolver_aws_cloudmap.go
Comment (jpkrohling) [reply exporter/loadbalancingexporter/resolver_aws_cloudmap.go:41]: Is there a way to know whether there are more than 100 hosts available? If so, perhaps it could warn a message in the logs?

**Review by jpkrohling — COMMENTED** [review]
Before this gets merged, I'd like to get clarification about the health status filter. That being clarified, this can be merged.

**Review by jpkrohling — CHANGES_REQUESTED** [review]
(changing to "request changes", to reflect the last comment)

**Comment by andretong** [comment]
@jpkrohling Thanks for the feedback.

> I might be missing something, but the value the user is providing is indeed being used, and therefore, deserves to be documented.

You're right, I've added the `healthStatus` documentation to the readme.

> Is there a way to know whether there are more than 100 hosts available? If so, perhaps it could warn a message in the logs?

With the current implementation will return at most 100 hosts using the [DiscoverInstances](https://docs.aws.amazon.com/cloud-map/latest/api/API_DiscoverInstances.html#API_DiscoverInstances_RequestSyntax) method.

Checking on the [documentation](https://docs.aws.amazon.com/cloud-map/latest/api/API_ListInstances.html), we could retrieve if we have more than 100 hosts by using `ListInstances` method but this change would be a major change that I think is best covered in this [feature request](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/29771), WDYT?

On the other hand, I've set this note in the readme to at least inform the users about this limitation

**Review by jpkrohling — APPROVED** [review]

**Comment by andretong** [comment]
@jpkrohling sorry for the delay.
I've rebased from `main` and fixed checks.

**Comment by andretong** [comment]
@jpkrohling rebased again from main.
Regarding the follow-up process, let me know if I'm missing any steps to move forward.
Thanks

**Comment by andretong** [comment]
@jpkrohling regarding the merge, should I add myself as a CODEOWNER and proceed with the merge?
I've check the documentation sections like [this](https://github.com/open-telemetry/opentelemetry-collector-contrib?tab=readme-ov-file#prs-and-reviews) and [this](https://github.com/open-telemetry/community/blob/main/CONTRIBUTING.md) but didn't find that particular procedure.

Thanks

**Comment by jpkrohling** [comment]
@andretong, this is ready to be merged, and as soon as the CI is green, I'll merge it. Feel free to ping me once this gets green, but I'll try to follow this closely.

**Comment by andretong** [comment]
> @andretong, this is ready to be merged, and as soon as the CI is green, I'll merge it. Feel free to ping me once this gets green, but I'll try to follow this closely.

Thanks @jpkrohling, I've rebased once more from `main`.
Not sure what cause the issue in the previous rebase, but it should be green already in this one 🤞

**Comment by jpkrohling** [comment]
Apparently, our main is currently broken. The PR is failing due to an unrelated change. Once our main is fixed, I'll rebase this.

**Comment by jpkrohling** [comment]
I think there might be a compilation issue caused by the rebases. Can you take a look and update the PR?

**Comment by andretong** [comment]
> I think there might be a compilation issue caused by the rebases. Can you take a look and update the PR?

Sure, rebased. thanks

**Comment by jpkrohling** [comment]
For some reason, CI is still failing. Can you figure out what's going on?

**Comment by andretong** [comment]
> For some reason, CI is still failing. Can you figure out what's going on?

It looks like it was a missing entry at the`go.sum` without the aws-sdk.
I've rebased once more

**Comment by jpkrohling** [comment]
Another failure :-/ 

```
Error: exporter/awskinesisexporter/exporter.go:10:2: missing go.sum entry for module providing package github.com/aws/aws-sdk-go-v2/aws (imported by github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awskinesisexporter); to add:
	go get github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awskinesisexporter@v0.96.0
```

**Comment by andretong** [comment]
> Another failure :-/
> 
> ```
> Error: exporter/awskinesisexporter/exporter.go:10:2: missing go.sum entry for module providing package github.com/aws/aws-sdk-go-v2/aws (imported by github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awskinesisexporter); to add:
> 	go get github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awskinesisexporter@v0.96.0
> ```

@jpkrohling I think I've fixed the errors, can you run the pipeline please?

I've check locally and the distribution is built when running `make otelcontribcol`

**Comment by andretong** [comment]
image
@jpkrohling I've uploaded the missing artifact in the last commit.

**Comment by andretong** [comment]
@jpkrohling seems is OK but a need rebase was needed :( 
Screenshot 2024-03-13 at 14 16 50

**Comment by mx-psi** [comment]
Merging based on https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/27588#issuecomment-1971641072 cc @jpkrohling

**Pull request merged**
